dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/ApacheHttpClientBlockingChannel.java

```diff
@@ -115,6 +116,23 @@ public Response execute(Endpoint endpoint, Request request) throws IOException {
             // not retried by default, so ours implements SafeLoggable and retains the simple-name for
             // cleaner metrics.
             throw new SafeConnectTimeoutException(e, failureDiagnosticArgs(endpoint, request, startTime));
+        } catch (NoHttpResponseException e) {
+            // NoHttpResponseException may be thrown immediately when a request is sent if a pooled persistent
+            // connection has been closed by the target server, or an intermediate proxy. In this case it's
+            // important that we retry the request with a fresh connection.
+            // The other possibility is that a remote server or proxy may time out an active request due
+            // to inactivity and close the connection without a response, in this case the request mustn't
+            // be retried.
+            // We attempt to differentiate these two cases based on request duration, we expect most of
+            // the prior case to occur within a couple milliseconds, however we must use a larger value
+            // to account for large garbage collections.
+            long durationNanos = System.nanoTime() - startTime;
+            Arg<?>[] diagnosticArgs = failureDiagnosticArgs(endpoint, request, startTime);
+            if (durationNanos < TimeUnit.SECONDS.toNanos(5)) {
+                e.addSuppressed(new Diagnostic(diagnosticArgs));
```

Comment:
So if it's below 5 seconds, RetryingChannel will retry the request.
If it's longer than 5 seconds, it goes through the SocketTimeoutException codepath, and that only retries "connect timed out" exceptions.